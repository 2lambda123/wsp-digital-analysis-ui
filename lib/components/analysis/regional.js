// @flow
/** render a regional analysis */

import Icon from '@conveyal/woonerf/components/icon'
import React, {Component} from 'react'
import Select from 'react-select'
import {sprintf} from 'sprintf-js'

import AggregateAccessibility from './aggregate-accessibility'
import Legend from './legend'
import colors from '../../constants/colors'
import messages from '../../utils/messages'
import ProfileRequestDisplay from './profile-request-display'
import Collapsible from '../collapsible'
import {Slider, Group, Text, File} from '../input'

import type {
  RegionalAnalysis,
  Grid,
  AggregationArea,
  AggregateAccessibility as AggregateAccessibilityType,
  Indicator,
  ReactSelectResult
} from '../../types'

type Props = {
  analysis: ?RegionalAnalysis, // can be null when loading
  analysisId: string,
  breaks: number[],
  comparisonAnalysis: ?RegionalAnalysis,
  differenceGrid: ?Grid,
  grid: ?Grid,
  aggregationAreas: AggregationArea[],
  indicators: Indicator[],
  minimumImprovementProbability: number,
  projectId: string,
  regionalAnalyses: ?(RegionalAnalysis[]),
  aggregationAreaId: ?string,
  aggregationWeightsId: ?string,
  aggregationAreaUploading: ?boolean,
  aggregateAccessibility: ?AggregateAccessibilityType,
  comparisonAggregateAccessibility: ?AggregateAccessibilityType,
  accessToName: string,
  comparisonAccessToName: ?string,
  aggregationWeightsName: ?string,

  // actions
  addRegionalAnalysisLayerToMap: () => void,
  addRegionalComparisonLayerToMap: () => void,
  fetch: any => any,
  loadRegionalAnalyses: string => void,
  removeRegionalAnalysisLayerFromMap: () => void,
  removeRegionalComparisonLayerFromMap: () => void,
  setActiveRegionalAnalysis: any => void,
  setMinimumImprovementProbability: number => void,
  setAggregationArea: ({aggregationAreaId: string, projectId: string}) => void,
  setAggregationWeights: ({
    aggregationWeightsId: string,
    projectId: string
  }) => void,
  uploadAggregationArea: (arg: {
    name: string,
    files: FileList,
    projectId: string
  }) => void,
  addAggregationAreaComponentToMap: () => void,
  removeAggregationAreaComponentFromMap: () => void
}

type State = {
  showAggregationAreaUpload: boolean,
  aggregationAreaFiles: null | FileList,
  aggregationAreaName: null | string
}

export default class RegionalAnalysisDisplay extends Component<
  void,
  Props,
  State
> {
  state = {
    aggregationAreaFiles: null,
    aggregationAreaName: null,
    showAggregationAreaUpload: false
  }

  componentWillMount () {
    const {
      analysis,
      analysisId,
      setActiveRegionalAnalysis,
      addRegionalAnalysisLayerToMap,
      loadRegionalAnalyses,
      projectId
    } = this.props
    if (analysis == null) {
      loadRegionalAnalyses(projectId)
    }

    setActiveRegionalAnalysis({id: analysisId})
    addRegionalAnalysisLayerToMap()
  }

  selectComparisonAnalysis = (chosen: ReactSelectResult): void => {
    const {setActiveRegionalAnalysis, analysisId} = this.props

    if (chosen != null) {
      setActiveRegionalAnalysis({id: analysisId, comparisonId: chosen.value})
    } else {
      // not a comparison
      setActiveRegionalAnalysis({id: analysisId})
    }
  }

  selectAggregationArea = (e: ReactSelectResult) => {
    const {
      projectId,
      setAggregationArea,
      addAggregationAreaComponentToMap,
      removeAggregationAreaComponentFromMap
    } = this.props

    if (e) {
      if (e.value === 'new-aggregation-area') {
        this.setState({showAggregationAreaUpload: true})
      } else {
        setAggregationArea({projectId, aggregationAreaId: e.value})
        if (e.value != null) addAggregationAreaComponentToMap()
        else removeAggregationAreaComponentFromMap()
      }
    }
  }

  selectAggregationWeights = (e: ReactSelectResult) => {
    if (e) {
      const {projectId, setAggregationWeights} = this.props
      setAggregationWeights({projectId, aggregationWeightsId: e.value})
    }
  }

  renderComparisonSelector () {
    const {regionalAnalyses, comparisonAnalysis, analysis} = this.props

    if (!analysis || !regionalAnalyses) return null // don't render until loaded

    const options = regionalAnalyses
      .filter(({id}) => id !== analysis.id)
      // don't render deleted analyses
      .filter(({deleted}) => !deleted)
      // don't compare incomparable analyses
      .filter(
        ({zoom, width, height, north, west}) =>
          zoom === analysis.zoom &&
          width === analysis.width &&
          height === analysis.height &&
          north === analysis.north &&
          west === analysis.west
      )
      .map(({id, name}) => {
        return {value: id, label: name}
      })

    return (
      <div>
        <Select
          value={comparisonAnalysis ? comparisonAnalysis.id : null}
          options={options}
          onChange={this.selectComparisonAnalysis}
          placeholder={messages.analysis.compareTo}
        />
      </div>
    )
  }

  /** Called when editing the new aggregation area name */
  changeAggregationAreaName = (e: Event & {target: {value: string}}): void => {
    this.setState({aggregationAreaName: e.target.value})
  }

  /** Called when selecting files to upload */
  changeAggregationAreaFiles = (e: Event & {target: {files: File[]}}): void => {
    this.setState({aggregationAreaFiles: e.target.files})
  }

  uploadAggregationArea = (e: Event): void => {
    e.preventDefault()
    const {uploadAggregationArea, projectId} = this.props
    const {aggregationAreaFiles, aggregationAreaName} = this.state

    if (!aggregationAreaName || !aggregationAreaFiles) {
      throw new Error('Attempt to upload null aggregation area')
    }

    uploadAggregationArea({
      name: aggregationAreaName,
      files: aggregationAreaFiles,
      projectId
    })
    this.setState({
      aggregationAreaName: null,
      aggregationAreaFiles: null,
      showAggregationAreaUpload: false
    })
  }

  /** Render controls for weighted average accessibility */
  renderAggregation () {
    const {
      aggregationAreas,
      aggregationAreaId,
      indicators,
      aggregationWeightsId,
      aggregateAccessibility,
      comparisonAggregateAccessibility,
      comparisonAnalysis,
      aggregationAreaUploading,
      analysis
    } = this.props
    const {
      showAggregationAreaUpload,
      aggregationAreaName,
      aggregationAreaFiles
    } = this.state

    // this same conditional is in the render function before this function is ever called, but flow
    // doesn't know that
    if (!analysis) return

    const aggregationAreaOptions = aggregationAreas.map(aa => ({
      value: aa.id,
      label: aa.name
    }))
    aggregationAreaOptions.push({
      label: messages.analysis.newAggregationArea,
      value: 'new-aggregation-area' // IDs of aggregation areas are UUIDs so this will not conflict
    })
    const opportunityDatasetOptions = indicators.map(od => ({
      value: od.key,
      label: od.name
    }))

    return (
      <div>
        <Group label={messages.analysis.aggregateTo}>
          <Select
            name='aggregateTo'
            options={aggregationAreaOptions}
            value={aggregationAreaId}
            onChange={this.selectAggregationArea}
          />
        </Group>

        {showAggregationAreaUpload &&
          <form onSubmit={this.uploadAggregationArea}>
            <Group label={messages.analysis.aggregationAreaName}>
              <Text
                name='Name'
                value={aggregationAreaName}
                onChange={this.changeAggregationAreaName}
              />
            </Group>

            <Group label={messages.analysis.aggregationAreaFiles}>
              <File
                multiple
                name='files'
                onChange={this.changeAggregationAreaFiles}
              />
            </Group>

            <input
              className='btn btn-block btn-success'
              disabled={
                aggregationAreaUploading ||
                !aggregationAreaName ||
                !aggregationAreaFiles
              }
              type='submit'
              value={messages.analysis.uploadAggregationArea}
            />
          </form>}

        {aggregationAreaId &&
          <Group label={messages.analysis.weightBy}>
            <Select
              name='aggregationWeights'
              options={opportunityDatasetOptions}
              value={aggregationWeightsId}
              onChange={this.selectAggregationWeights}
            />
          </Group>}

        {aggregationAreaUploading &&
          <span>
            <Icon type='fa-spinner' spin />&nbsp;{messages.analysis.uploading}
          </span>}

        {aggregateAccessibility &&
          aggregationWeightsId &&
          <AggregateAccessibility
            aggregateAccessibility={aggregateAccessibility}
            comparisonAggregateAccessibility={comparisonAggregateAccessibility}
            weightByName={this.findIndicatorName(aggregationWeightsId)}
            accessToName={this.findIndicatorName(analysis.grid)}
            regionalAnalysisName={analysis.name}
            comparisonAccessToName={
              comparisonAnalysis &&
              this.findIndicatorName(comparisonAnalysis.grid)
            }
            comparisonRegionalAnalysisName={
              comparisonAnalysis && comparisonAnalysis.name
            }
          />}
      </div>
    )
  }

  /**
   * Perform an authenticated fetch to get a presigned URL to download a grid from S3,
   * then download it. Pass in the path to fetch.
   */
  downloadFromS3 (url: string) {
    const {fetch} = this.props
    fetch({
      url,
      next (error, response) {
        if (error) {
          window.alert(error)
        } else {
          window.open(response.value.url)
        }
      }
    })
  }

  downloadScenarioGIS = (e: Event) => {
    e.preventDefault()
    if (this.props.analysis) {
      this.downloadFromS3(
        `/api/regional/${this.props.analysis.id}/grid/tiff?redirect=false`
      )
    } else {
      throw new Error('Attempt to dowload analysis that has not yet loaded!')
    }
  }

  downloadComparisonGIS = (e: Event) => {
    e.preventDefault()
    if (this.props.comparisonAnalysis) {
      this.downloadFromS3(
        `/api/regional/${this.props.comparisonAnalysis
          .id}/grid/tiff?redirect=false`
      )
    } else {
      throw new Error(
        'Attempt to dowload comparison analysis that has not yet loaded!'
      )
    }
  }

  downloadProbabilityGIS = (e: Event) => {
    e.preventDefault()
    // NB base comes first (TODO change?)
    if (this.props.analysis && this.props.comparisonAnalysis) {
      this.downloadFromS3(
        `/api/regional/${this.props.comparisonAnalysis.id}/${this.props.analysis
          .id}/tiff?redirect=false`
      )
    } else {
      throw new Error(
        'Attempt to download probability surface when both analyses have not loaded.'
      )
    }
  }

  renderGisExport () {
    const {analysis, comparisonAnalysis} = this.props

    // this is also done in the main render block but flow doesn't know that
    if (!analysis) return

    // we use <a> tags not buttons in order to get text wrapping; all event handlers prevent default
    return (
      <div>
        <b>
          <Icon type='map-o' />
          {messages.analysis.gisExport}
        </b>
        <ul>
          <li>
            <a onClick={this.downloadScenarioGIS} tabIndex={0}>
              {sprintf(messages.analysis.downloadRegional, analysis.name)}
            </a>
          </li>
          {comparisonAnalysis &&
            <li>
              <a onClick={this.downloadComparisonGIS} tabIndex={0}>
                {sprintf(
                  messages.analysis.downloadRegional,
                  comparisonAnalysis.name
                )}
              </a>
            </li>}
          {comparisonAnalysis &&
            <li>
              <a onClick={this.downloadProbabilityGIS} tabIndex={0}>
                {messages.analysis.downloadRegionalProbability}
              </a>
            </li>}
        </ul>
      </div>
    )
  }

  render () {
    const {
      analysis,
      breaks,
      comparisonAnalysis,
      differenceGrid,
      grid,
      indicators,
      minimumImprovementProbability,
      setMinimumImprovementProbability
    } = this.props

    if (analysis == null || grid == null || indicators == null) return null

    return (
      <div>
        <h3>
          {analysis.name}
        </h3>

        {this.renderComparisonSelector()}

        {comparisonAnalysis &&
          <MinimumImprovementProbabilitySelector
            minimumImprovementProbability={minimumImprovementProbability}
            setMinimumImprovementProbability={setMinimumImprovementProbability}
          />}

        <div>
          <i>
            {analysis.travelTimePercentile !== -1 &&
              sprintf(messages.analysis.accessTo, {
                opportunity: this.findIndicatorName(analysis.grid),
                cutoff: analysis.cutoffMinutes,
                percentile: analysis.travelTimePercentile
              })}
            {analysis.travelTimePercentile === -1 &&
              sprintf(messages.analysis.accessToInstantaneous, {
                opportunity: this.findIndicatorName(analysis.grid),
                cutoff: analysis.cutoffMinutes
              })}
          </i>
        </div>

        {comparisonAnalysis &&
          <div>
            <i>
              {comparisonAnalysis.travelTimePercentile !== -1 &&
                sprintf(messages.analysis.comparisonAccessTo, {
                  opportunity: this.findIndicatorName(comparisonAnalysis.grid),
                  cutoff: comparisonAnalysis.cutoffMinutes,
                  percentile: comparisonAnalysis.travelTimePercentile
                })}
              {comparisonAnalysis.travelTimePercentile === -1 &&
                sprintf(messages.analysis.comparisonAccessToInstantaneous, {
                  opportunity: this.findIndicatorName(comparisonAnalysis.grid),
                  cutoff: comparisonAnalysis.cutoffMinutes
                })}
            </i>
          </div>}

        <Legend
          breaks={breaks}
          min={
            comparisonAnalysis != null && differenceGrid != null
              ? differenceGrid.min
              : grid.min
          }
          colors={
            comparisonAnalysis != null
              ? colors.REGIONAL_COMPARISON_GRADIENT
              : colors.REGIONAL_GRADIENT
          }
        />

        <Collapsible
          title={sprintf(messages.analysis.settingsFor, analysis.name)}
        >
          <ProfileRequestDisplay request={analysis.request} />
        </Collapsible>

        {comparisonAnalysis &&
          <Collapsible
            title={sprintf(
              messages.analysis.settingsFor,
              comparisonAnalysis.name
            )}
          >
            <ProfileRequestDisplay request={comparisonAnalysis.request} />
          </Collapsible>}

        {this.renderAggregation()}

        {this.renderGisExport()}
      </div>
    )
  }

  componentWillUnmount () {
    const {
      removeRegionalAnalysisLayerFromMap,
      removeRegionalComparisonLayerFromMap
    } = this.props
    removeRegionalAnalysisLayerFromMap()
    removeRegionalComparisonLayerFromMap()
  }

  // TODO replace with selector
  findIndicatorName (key: string): string {
    const value = this.props.indicators.find(i => i.key === key)
    return value != null ? value.name : key
  }
}

function MinimumImprovementProbabilitySelector ({
  minimumImprovementProbability,
  setMinimumImprovementProbability
}) {
  return (
    <Slider
      label={messages.analysis.minimumImprovementProbability}
      output
      format='.2f'
      min={0}
      max={1}
      step={0.01}
      value={minimumImprovementProbability}
      onChange={e =>
        setMinimumImprovementProbability(parseFloat(e.target.value))}
    />
  )
}
