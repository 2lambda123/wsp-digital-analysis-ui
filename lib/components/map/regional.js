import React, { Component, PropTypes } from 'react'
import { ReactChoropleth } from '@conveyal/gridualizer'
import lonlat from '@conveyal/lonlat'
import colors from '../../constants/colors'

const NO_DATA = -Math.pow(2, 30)

/** A layer showing a proabilistic regional comparison */
export default class Regional extends Component {
  static propTypes = {
    baseGrid: PropTypes.object.isRequired,
    differenceGrid: PropTypes.object.isRequired,
    probabilityGrid: PropTypes.object.isRequired,
    minimumImprovementProbability: PropTypes.number,
    breaks: PropTypes.array.isRequired,
    id: PropTypes.string.isRequired,
    comparisonId: PropTypes.string
  }

  static defaultProps = {
    minimumImprovementProbability: 0.95 // https://xkcd.com/1478/
  }

  static contextTypes = {
    map: PropTypes.object.isRequired
  }

  componentDidMount () {
    // TODO clooge, handle click actions the way we handle map state
    // maybe just call onClick on every map component if it has an onClick function
    const {map} = this.context
    map.on('click', this.onMapClick)
  }

  componentWillUnmount () {
    const {isRegionalAnalysisSamplingDistributionComponentOnMap, removeRegionalAnalysisSamplingDistributionComponentFromMap} = this.props
    const {map} = this.context
    map.off('click', this.onMapClick)
    if (isRegionalAnalysisSamplingDistributionComponentOnMap) {
      // don't leave dest travel time distribution on map when exiting analysis mode
      removeRegionalAnalysisSamplingDistributionComponentFromMap()
    }
  }

  onMapClick = (e) => {
    const {id, comparisonId, setRegionalAnalysisOrigin, isRegionalAnalysisSamplingDistributionComponentOnMap} = this.props
    // if the destination component is already on the map, don't add it - change destination by
    // dragging destination marker
    if (!isRegionalAnalysisSamplingDistributionComponentOnMap) {
      setRegionalAnalysisOrigin({ regionalAnalysisId: id, comparisonRegionalAnalysisId: comparisonId, latlon: lonlat(e.latlng) })
    }
  }

  render () {
    const { baseGrid, comparisonId, differenceGrid, probabilityGrid, minimumImprovementProbability, breaks } = this.props

    if (comparisonId == null) {
      return <ReactChoropleth grid={baseGrid} breaks={breaks} colors={colors.REGIONAL_GRADIENT} />
    } else {
      // now, mask with probability; probability is put in grids as ints 0 - 100k
      const maskedGrid = mask(differenceGrid, probabilityGrid, minimumImprovementProbability * 1e5)

      return <ReactChoropleth
        grid={maskedGrid}
        colors={colors.REGIONAL_COMPARISON_GRADIENT}
        breaks={breaks}
        noDataValue={NO_DATA}
        />
    }
  }
}

// mask a grid, returning a new, masked grid
function mask (grid, probabilityGrid, cutoff) {
  const gridsDoNotAlign = grid.west !== probabilityGrid.west ||
    grid.north !== probabilityGrid.north ||
    grid.zoom !== probabilityGrid.zoom ||
    grid.width !== probabilityGrid.width ||
    grid.height !== probabilityGrid.height

  if (gridsDoNotAlign) {
    throw new Error('Grids do not align for masking')
  }

  const { north, west, zoom, width, height } = grid

  const newGrid = {
    north,
    west,
    zoom,
    width,
    height,
    data: new Int32Array(width * height),
    // start at zero not Infinity to avoid having to update min/max for cells where probability
    // is less than the cutoff
    min: 0,
    max: 0
  }

  for (let pixel = 0; pixel < width * height; pixel++) {
    if (probabilityGrid.data[pixel] < cutoff) newGrid.data[pixel] = 0
    else {
      newGrid.data[pixel] = grid.data[pixel]
      if (newGrid.min > grid.data[pixel]) newGrid.min = grid.data[pixel]
      if (newGrid.max < grid.data[pixel]) newGrid.max = grid.data[pixel]
    }
  }

  return newGrid
}
