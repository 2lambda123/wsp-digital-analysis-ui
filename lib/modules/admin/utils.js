// @flow
import {connect} from 'react-redux'
import {createStructuredSelector} from 'reselect'

import * as actions from './actions'
import * as select from './selectors'

export function fullyConnect (Component: any) {
  return connect(
    createStructuredSelector(select),
    actions
  )(Component)
}
