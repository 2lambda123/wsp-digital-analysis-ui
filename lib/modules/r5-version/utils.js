//
import {VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version) {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || 0) / 100
    )
  }

  return 0
}
