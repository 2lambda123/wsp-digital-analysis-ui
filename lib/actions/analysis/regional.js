// @flow
/** Actions for regional analysis */

import {createGrid} from 'browsochrones'
import {classifiers} from '@conveyal/gridualizer'
import fetch, {fetchMultiple} from '@conveyal/woonerf/fetch'
import sortBy from 'lodash/sortBy'
import {createAction} from 'redux-actions'
import {push} from 'react-router-redux'

import colors from '../../constants/colors'
import {
  REGIONAL_ANALYSIS_SAMPLING_DISTRIBUTION_COMPONENT
} from '../../constants/map'
import {addComponent} from '../map'
import selectCurrentProjectId from '../../selectors/current-project-id'

import OpportunityDatasets from '../../modules/opportunity-datasets'
import R5Version from '../../modules/r5-version'

import type {Bounds, LonLat} from '../../types'

const REGIONAL_URL = '/api/regional'

export const setRegionalAnalyses = createAction('set regional analyses')

export const load = (regionId: string) =>
  fetch({
    url: `/api/region/${regionId}/regional`,
    next (response) {
      const analyses = sortBy(response.value, a => -a.createdAt) // newest at the top
      return [
        setRegionalAnalyses(response.value),
        R5Version.actions.setLastAnalysisVersion(analyses.length > 0 ? analyses[0].workerVersion : undefined)
      ]
    }
  })

export const setActiveRegionalAnalyses = createAction(
  'set active regional analyses'
)
export const setRegionalAnalysisGrids = createAction(
  'set regional analysis grids'
)
export const setMinimumImprovementProbability = createAction(
  'set minimum improvement probability'
)
export const setRegionalAnalysisBounds = createAction(
  'set regional analysis bounds'
)
const setRegionalAnalysisOriginLocally = createAction(
  'set regional analysis origin'
)
const setRegionalAnalysisSamplingDistribution = createAction(
  'set regional analysis sampling distribution'
)
const setComparisonRegionalAnalysisSamplingDistribution = createAction(
  'set comparison regional analysis sampling distribution'
)

/** This also does not add anything to the map */
export const loadRegionalAnalysisGrids = async ({
  _id,
  comparisonId
}: {
  _id: string,
  comparisonId: string
}) => {
  if (comparisonId) {
    return [
      setActiveRegionalAnalyses({
        _id,
        comparisonId
      }),
      fetchMultiple({
        fetches: [
          `${REGIONAL_URL}/${_id}/grid/grid`,
          `${REGIONAL_URL}/${comparisonId}/grid/grid`,
          `${REGIONAL_URL}/${comparisonId}/${_id}/grid`
        ].map(url => ({url})),
        next ([gridResponse, comparisonGridResponse, probabilityGridResponse]) {
          // TODO store raw data, create grids, classifier, breaks in selectors
          const comparisonGrid = createGrid(comparisonGridResponse.value)
          const grid = createGrid(comparisonGridResponse.value)
          const probabilityGrid = createGrid(probabilityGridResponse.value)
          const differenceGrid = subtract(grid, comparisonGrid)
          const classifier = classifiers.ckmeans({})
          const breaks = classifier(
            differenceGrid,
            colors.REGIONAL_COMPARISON_GRADIENT.length
          )

          return setRegionalAnalysisGrids({
            classifier,
            grid,
            comparisonGrid,
            probabilityGrid,
            differenceGrid,
            breaks
          })
        }
      })
    ]
  } else {
    return [
      setActiveRegionalAnalyses({_id, comparisonId}),
      fetch({
        url: `${REGIONAL_URL}/${_id}/grid/grid`,
        next (response) {
          const gridData = response.value
          const grid = createGrid(gridData)
          const classifier = classifiers.ckmeans({})
          const breaks = classifier(grid, colors.REGIONAL_GRADIENT.length)
          return setRegionalAnalysisGrids({
            classifier,
            grid,
            breaks
          })
        }
      })
    ]
  }
}

export const createRegionalAnalysis = ({
  bounds,
  name,
  profileRequest,
  travelTimePercentile,
  variantId
}: {
  bounds: Bounds,
  name: string,
  profileRequest: any,
  travelTimePercentile: number,
  variantId: number
}) => (dispatch: () => void, getState: () => any) => {
  const state = getState()
  const currentProjectId = selectCurrentProjectId(state, {})
  const opportunityDataset = OpportunityDatasets.selectors.selectActiveOpportunityDataset(state, {})
  const workerVersion = R5Version.selectors.selectCurrentR5Version(state, {})

  if (!opportunityDataset || !opportunityDataset.key) {
    window.alert('Opportunity dataset must be selected.')
    return
  }

  dispatch(
    fetch({
      options: {
        method: 'POST',
        body: {
          ...profileRequest,
          bounds,
          maxTripDurationMinutes: state.analysis.isochroneCutoff,
          name,
          opportunityDatasetKey: opportunityDataset.key,
          projectId: currentProjectId,
          travelTimePercentile,
          variantIndex: variantId,
          workerVersion
        }
      },
      url: REGIONAL_URL,
      next: () => push(`/projects/${currentProjectId}/regional`)
    })
  )
}

const deleteRegionalAnalysisLocally = createAction('delete regional analysis')
export const deleteRegionalAnalysis = (analysisId: string) => [
  // run local delete first so it seems snappier. The worst that will happen is
  // that a regional analysis will pop back up in a few seconds when we refresh
  // regional analyses
  deleteRegionalAnalysisLocally(analysisId),
  fetch({
    url: `${REGIONAL_URL}/${analysisId}`,
    options: {
      method: 'DELETE'
    }
  })
]

const setAggregationAreaLocally = createAction('set aggregation area')
const setAggregationAreaId = createAction('set aggregation area id')
export const setAggregationArea = async ({
  regionId,
  aggregationAreaId
}: {
  regionId: string,
  aggregationAreaId: string
}) => [
  setAggregationAreaLocally(null), // clear existing so UI doesn't show results that don't match now-selected aggregation area
  setAggregationAreaId(aggregationAreaId),
  fetch({
    url: `/api/region/${regionId}/aggregationArea/${aggregationAreaId}`,
    next: (res) => setAggregationAreaLocally(createGrid(res.value))
  })
]

const setAggregationWeightsLocally = createAction('set aggregation weights')
const setAggregationWeightsId = createAction('set aggregation weights id')
export const setAggregationWeights = ({
  regionId,
  aggregationWeightsId
}: {
  regionId: string,
  aggregationWeightsId: string
}) => [
  // clear existing so UI doesn't show results that don't match now-selected aggregaton weights
  setAggregationWeightsLocally(null),
  setAggregationWeightsId(aggregationWeightsId),
  fetch({
    url: `/api/grid/${regionId}/${aggregationWeightsId}`,
    next: (res) => setAggregationWeightsLocally(createGrid(res.value))
  })
]

const setAggregationAreaUploading = createAction(
  'set aggregation area uploading'
)
/** This triggers a reducer in region to inject the aggregation area into the current region */
const addAggregationAreaLocally = createAction('add aggregation area locally')

/** Upload an aggregation area */
export const uploadAggregationArea = ({
  name,
  files,
  regionId
}: {
  name: string,
  files: FileList,
  regionId: string
}) => {
  const formData = new window.FormData()
  formData.append('name', name)
  ;[...files].forEach(file => formData.append('files', file))

  return [
    setAggregationAreaUploading(true),
    fetch({
      url: `/api/region/${regionId}/aggregationArea`,
      options: {
        method: 'post',
        body: formData
      },
      next: (res) => [
        addAggregationAreaLocally(res.value),
        setAggregationAreaUploading(false)
      ]
    })
  ]
}

/** non-destructively subtract grid B from grid A */
function subtract (a, b) {
  const gridsDoNotAlign =
    a.west !== b.west ||
    a.north !== b.north ||
    a.zoom !== b.zoom ||
    a.width !== b.width ||
    a.height !== b.height

  if (gridsDoNotAlign) {
    throw new Error('Grids do not align for subtraction')
  }

  const {north, west, zoom, width, height} = a

  const newGrid = {
    north,
    west,
    zoom,
    width,
    height,
    data: new Int32Array(width * height),
    min: Infinity,
    max: -Infinity
  }

  for (let pixel = 0; pixel < width * height; pixel++) {
    const val = a.data[pixel] - b.data[pixel]
    newGrid.min = Math.min(newGrid.min, val)
    newGrid.max = Math.max(newGrid.max, val)
    newGrid.data[pixel] = val
  }

  return newGrid
}

export function downloadGridFromS3 (url: string) {
  return fetch({
    url,
    next: (err, response) => {
      if (err) window.alert(err)
      else window.open(response.url)
    }
  })
}

/** set the origin for showing a bootstrapped sampling distribution of travel time */
export const setRegionalAnalysisOrigin = ({
  comparisonRegionalAnalysisId,
  lonlat,
  regionalAnalysisId
}: {
  comparisonRegionalAnalysisId: string,
  lonlat: LonLat,
  regionalAnalysisId: string
}) => {
  const fetches = [
    {
      url: `${REGIONAL_URL}/${regionalAnalysisId}/samplingDistribution/${lonlat.lat}/${lonlat.lon}`
    }
  ]

  if (comparisonRegionalAnalysisId) {
    fetches.push({
      url: `${REGIONAL_URL}/${comparisonRegionalAnalysisId}/samplingDistribution/${lonlat.lat}/${lonlat.lon}`
    })
  }

  return fetchMultiple({
    fetches,
    next: (responses) => [
      setRegionalAnalysisOriginLocally(lonlat),
      setRegionalAnalysisSamplingDistribution(responses[0].value),
      setComparisonRegionalAnalysisSamplingDistribution(
        comparisonRegionalAnalysisId ? responses[1].value : null
      ),
      addComponent(REGIONAL_ANALYSIS_SAMPLING_DISTRIBUTION_COMPONENT)
    ]
  })
}
