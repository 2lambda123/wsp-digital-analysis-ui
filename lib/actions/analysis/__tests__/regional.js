/* global describe, expect, it */

import * as regional from '../regional'
import nock from 'nock'

import { mockRegionalAnalyses } from '../../../utils/mock-data'

describe('Actions > Analysis > Regional', () => {
  it('setRegionalAnalyses should work', () => {
    expect(regional.setRegionalAnalyses(mockRegionalAnalyses)).toMatchSnapshot()
  })

  it('setActiveRegionalAnalyses should work', () => {
    expect(regional.setActiveRegionalAnalyses()).toMatchSnapshot()
  })

  it('setRegionalAnalysisGrids should work', () => {
    expect(regional.setRegionalAnalysisGrids()).toMatchSnapshot()
  })

  it('setMinimumImprovementProbability should work', () => {
    expect(regional.setMinimumImprovementProbability(0.42)).toMatchSnapshot()
  })

  it('deleteRegionalAnalysis should work', async () => {
    const nockHost = nock('http://mockhost.com')
      .delete('/api/regional/12321a')
      .reply(200, 'deleted')

    const [deleteLocally, deleteFromServer] = regional.deleteRegionalAnalysis('12321a')
    expect(deleteLocally).toMatchSnapshot()

    await deleteFromServer
    expect(nockHost.isDone()).toBeTruthy()
  })
  // TODO loadRegionalAnalysisGrids, runAnalysis, deleteRegionalAnalysis
})
