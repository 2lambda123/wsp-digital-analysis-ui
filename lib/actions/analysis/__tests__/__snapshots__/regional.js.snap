// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions > Analysis > Regional deleteRegionalAnalysis should work 1`] = `
Object {
  "payload": "12321a",
  "type": "delete regional analysis",
}
`;

exports[`Actions > Analysis > Regional setActiveRegionalAnalyses should work 1`] = `
Object {
  "type": "set active regional analyses",
}
`;

exports[`Actions > Analysis > Regional setMinimumImprovementProbability should work 1`] = `
Object {
  "payload": 0.42,
  "type": "set minimum improvement probability",
}
`;

exports[`Actions > Analysis > Regional setRegionalAnalyses should work 1`] = `
Object {
  "payload": Array [
    Object {
      "bundleId": "5678",
      "cutoffMinutes": 60,
      "grid": "Jobs_total",
      "height": 300,
      "id": "abcd",
      "name": "ANALYSIS 1",
      "north": 20024,
      "request": Object {
        "accessModes": "WALK",
        "bikeSafe": 1,
        "bikeSlope": 1,
        "bikeSpeed": 4.166666666666667,
        "bikeTime": 1,
        "bikeTrafficStress": 4,
        "carSpeed": 20,
        "date": "2016-01-16",
        "directModes": "WALK",
        "egressModes": "WALK",
        "fromTime": 25200,
        "maxBikeTime": 20,
        "maxCarTime": 45,
        "maxRides": 4,
        "maxWalkTime": 20,
        "minBikeTime": 10,
        "minCarTime": 10,
        "monteCarloDraws": 200,
        "reachabilityThreshold": 0,
        "streetTime": 90,
        "suboptimalMinutes": 5,
        "toTime": 32400,
        "transitModes": "BUS,TRAM,RAIL,SUBWAY",
        "walkSpeed": 1.3888888888888888,
      },
      "scenarioId": "1234",
      "travelTimePercentile": -1,
      "variant": 0,
      "west": 20015,
      "width": 300,
      "zoom": 9,
    },
    Object {
      "bundleId": "5678",
      "cutoffMinutes": 60,
      "grid": "Jobs_total",
      "height": 300,
      "id": "efgh",
      "name": "ANALYSIS 2",
      "north": 20024,
      "request": Object {
        "accessModes": "WALK",
        "bikeSafe": 1,
        "bikeSlope": 1,
        "bikeSpeed": 4.166666666666667,
        "bikeTime": 1,
        "bikeTrafficStress": 4,
        "carSpeed": 20,
        "date": "2016-01-16",
        "directModes": "WALK",
        "egressModes": "WALK",
        "fromTime": 25200,
        "maxBikeTime": 20,
        "maxCarTime": 45,
        "maxRides": 4,
        "maxWalkTime": 20,
        "minBikeTime": 10,
        "minCarTime": 10,
        "monteCarloDraws": 200,
        "reachabilityThreshold": 0,
        "streetTime": 90,
        "suboptimalMinutes": 5,
        "toTime": 32400,
        "transitModes": "BUS,TRAM,RAIL,SUBWAY",
        "walkSpeed": 1.3888888888888888,
      },
      "scenarioId": "4321",
      "travelTimePercentile": 42,
      "variant": 1,
      "west": 20015,
      "width": 300,
      "zoom": 9,
    },
  ],
  "type": "set regional analyses",
}
`;

exports[`Actions > Analysis > Regional setRegionalAnalysisGrids should work 1`] = `
Object {
  "type": "set regional analysis grids",
}
`;
